File: aoc2023/day12/day12.py
import math
import re
import time
from concurrent.futures import ProcessPoolExecutor, as_completed
from aocl import *


debug = False


def solve(input_file, expand, record_range=None):
    records = [splits(s) for s in read_lines(input_file)]

    for i, (springs, groups) in enumerate(records):
        springs = springs.replace('.', '_').replace('?', '.')
        groups = ints(groups)
        if expand > 1:
            springs = '.'.join([springs] * expand)
            groups *= expand
            records[i] = (springs, groups)
        records[i] = (springs, groups)

    start = time.time()
    total_arrangements = 0

    if record_range is not None:
        records = records[record_range[0]:record_range[1]]
    # unfinished = (303, 625, 925)
    unfinished = list(range(len(records)))
    total_finished = len(records) - len(unfinished)

    with ProcessPoolExecutor() as executor:
        future_to_idx = {
            executor.submit(begin_arrange, i, springs, groups): i
            for i, (springs, groups)
            in enumerate(records)
            if i in unfinished
        }
        for future in as_completed(future_to_idx):
            total_finished += 1
            idx = future_to_idx[future]
            try:
                num_arrangements = future.result()
            except Exception as exc:
                print('%r generated an exception: %s' % (idx, exc))
            else:
                total_arrangements += num_arrangements
                if debug:
                    springs, groups = records[idx]
                    print(f'r#{idx:04}:', springs)
                    print('  groups', groups)
                    print('  num arrangements:', num_arrangements, 'total now', total_arrangements)
                    print('  progress:', f'{math.floor(total_finished / len(records) * 100)}%',
                          f'({len(records)-total_finished} remaining)')

    if debug:
        print()
        print('total possible arrangements:', total_arrangements)
        print('time:', time.time() - start)
    return total_arrangements


def begin_arrange(idx, springs, groups):
    if debug:
        print(f'start r#{idx:04}')
    arrangement = [-1] * (len(groups)*2 + 1)
    return arrange(arrangement, springs, groups)


def arrange(arrangement, springs, groups, indent=4):
    # print(show_arrangement(arrangement), file=sys.stderr)
    gs = groups[0]
    ge = groups[-1]

    if min(arrangement) >= 0:
        return 0

    first_unset = arrangement.index(-1)
    last_unset = len(arrangement) - 1 - arrangement[::-1].index(-1)
    # print(' '*indent + 'arrange', last_unset - first_unset + 1, show_arrangement(arrangement), 'of', groups)

    gaps = len(groups) + 1
    min_p_len = sum(groups) + gaps
    if first_unset == 0:
        # no outer gaps needed
        min_p_len -= 2
        min_outside_space = 0
    else:
        min_outside_space = 1

    is_last = len(groups) <= 2
    is_single = len(groups) == 1

    used_before = sum(arrangement[:first_unset])
    used_after = sum(arrangement[last_unset + 1:])
    remaining_springs = springs[used_before:len(springs)-used_after]
    free_len = len(remaining_springs) - min_p_len
    # print(' '*indent + 'rs', remaining_springs, 'flen', free_len, 'minos', min_outside_space)
    if free_len < 0:
        # print(' '*indent + 'no 0, flen', free_len)
        return 0

    num_arrangements = 0
    if len(groups) > 2:
        inner_groups = groups[1:-1]
    else:
        inner_groups = None

    for s in range(min_outside_space, min_outside_space + free_len + 1):
        if 0 <= remaining_springs.find('#') < s:
            break
        test_s = '_' * s + '#' * gs
        if not re.match(remaining_springs[:len(test_s)], test_s):
            continue

        for e in range(min_outside_space, (min_outside_space + free_len + 1) - (s-min_outside_space)):
            if remaining_springs.rfind('#', len(remaining_springs)-e) >= 0:
                break
            test_e = '#' * ge + '_' * e

            if re.match(remaining_springs[-len(test_e):], test_e):
                if is_last:
                    if is_single and s + gs + e < len(remaining_springs):
                        # print(' '*indent + '=> no 1')
                        continue
                    elif remaining_springs.find('#', len(test_s), len(remaining_springs)-len(test_e)) >= 0:
                        # print(' '*indent + '=> no 2')
                        continue

                if is_single:  # (s, gs, e)
                    # arrangement[first_unset] = s
                    # arrangement[first_unset+1] = gs
                    # arrangement[last_unset] = e
                    # print(' '*indent + 'found single', (s, gs, e), num_arrangements + 1, show_arrangement(arrangement))
                    # arrangement[first_unset:last_unset+1] = -1
                    num_arrangements += 1
                elif is_last:  # (s, gs, len(remaining), ge, e)
                    # arrangement[first_unset] = s
                    # arrangement[first_unset+1] = gs
                    # arrangement[first_unset+2] = len(remaining)
                    # arrangement[last_unset-1] = gs
                    # arrangement[last_unset] = e
                    # print(' '*indent + 'found last', (s, gs, len(remaining), gs, e), num_arrangements + 1, show_arrangement(arrangement))
                    # arrangement[first_unset:last_unset+1] = -1
                    num_arrangements += 1
                elif inner_groups is None:  # (s, gs, ge, e)
                    num_arrangements += 1
                else:  # (s, gs, (subpattern), ge, e)
                    arrangement[first_unset] = s
                    arrangement[first_unset+1] = gs
                    arrangement[last_unset-1] = ge
                    arrangement[last_unset] = e
                    # print(' '*indent + 'potential match, checking inner groups')
                    num_arrangements += arrange(arrangement, springs, inner_groups, indent + 4)
                    arrangement[first_unset] = -1
                    arrangement[first_unset+1] = -1
                    arrangement[last_unset-1] = -1
                    arrangement[last_unset] = -1
            # else:
                # print(' '*indent + '=> no 3', test_s, remaining_springs[:len(test_s)], '*', test_e, remaining_springs[-len(test_e):], 'slen', len(remaining_springs))
    # print(' '*indent + 'arrange complete', num_arrangements)
    return num_arrangements


def show_arrangement(arrangement, width=0):
    def char(i, g):
        if i % 2 == 0:
            return '_' * g
        return '#' * g

    middle = len(arrangement) // 2
    start = ''.join([char(i, g) for i, g in enumerate(arrangement[:middle])])
    end = ''.join(reversed([char(i, g) for i, g in enumerate(reversed(arrangement[middle:]))]))
    if min(arrangement) < 0 and width > 0:
        return start + ' ' * (width - len(start) - len(end)) + end
    elif len(arrangement) % 2 == 0 or min(arrangement) < 0:
        return start + '|' + end
    else:
        return start + end


def main():
    _input_file = 'input'

    run(__file__, solve, _input_file, 7007, expand=0)
    run(__file__, solve, _input_file, 2113581, expand=3, record_range=(0, 201))


if __name__ == '__main__':
    main()

File: aoc2023/day12/java/src/Day12.java
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

import static java.util.Collections.min;

public class Day12 {
    public static void main(String[] args) {
        System.out.println("Running...");
        long start = System.nanoTime();

        String springs;
        List<Integer> groups;

        int expand = 5;
        AtomicLong sum = new AtomicLong( 0 );
        List<Arranger> arrangers = new ArrayList<Arranger>();

        // 303
        springs = expandS( "??????????????.", expand );
        groups = expandG( Arrays.asList( 1, 1, 1, 3, 1 ), expand );
        arrangers.add( new Arranger( "303", sum, springs, groups ) );

        // 625
        springs = expandS( "?.????????????.?", expand );
        groups = expandG( Arrays.asList( 1, 3, 1 ), expand );
        arrangers.add( new Arranger( "625", sum, springs, groups ) );

        // 925
        springs = expandS( "??..??..????????", expand );
        groups = expandG( Arrays.asList( 1, 1, 2 ), expand );
        arrangers.add( new Arranger( "925", sum, springs, groups ) );

        try {
            for (Arranger value : arrangers) {
                value.start();
            }
            System.out.println( "Started " + arrangers.size() + " threads" );

            for (Arranger arranger : arrangers) {
                arranger.join();
            }
        }
        catch (InterruptedException e) {
            System.out.println( "Interrupted?" );
            throw new RuntimeException(e);
        }

        System.out.println( "Sum: " + sum );

        float timeTaken = ( System.nanoTime() - start ) / 1000f / 1000f / 1000f;
        System.out.println( "Finished in " + timeTaken + "s" );
        if ( expand == 3 ) {
            assert sum.get() == 26712645;
        }
    }

    private static String expandS(String springs, int expand) {
        StringBuilder out = new StringBuilder();
        for (int i = 0; i < expand; i++) {
            out.append(springs);
            if (i < expand - 1) {
                out.append('?');
            }
        }
        return out.toString();
    }

    private static List<Integer> expandG(List<Integer> groups, int expand) {
        int newLength = groups.size() * expand;
        List<Integer> out = new ArrayList<>(newLength);
        for (int i = 0; i < newLength; i++) {
            out.add(groups.get(i % groups.size()));
        }
        return out;
    }

    private static class Arranger extends Thread {
        private final AtomicLong out;
        private final List<Integer> springs;
        private final List<Integer> groups;
        private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

        public Arranger(String name, AtomicLong out, String springs, List<Integer> groups) {
            setName( "t-arrange-" + name );

            this.out = out;
            this.groups = groups;

            List<Integer> springsList = new ArrayList<>(springs.length());
            for (int i = 0; i < springs.length(); i++) {
                char c = springs.charAt(i);
                if (c == '?') {
                    springsList.add(-1);
                }
                else if (c == '#') {
                    springsList.add(1);
                }
                else {
                    springsList.add(0);
                }
            }
            this.springs = springsList;
        }

        public void run() {
            int arrangementLength = groups.size() * 2 + 1;
            List<Integer> arrangement = new ArrayList<>(arrangementLength);
            for (int i = 0; i < arrangementLength; i++) {
                arrangement.add(-1);
            }
            long result = arrange(arrangement, springs, groups);
            long total = out.addAndGet( result );
            System.out.println( getName() + " finished, result = " + result + " total now = " + total );
        }

        private long arrange(List<Integer> arrangement, List<Integer> springs, List<Integer> groups) {
            int gs = groups.get(0);
            int ge = groups.get(groups.size() - 1);

            if (arrangement.isEmpty() || min(arrangement) >= 0) {
                return 0;
            }

            int first_unset = arrangement.indexOf(-1);
            int last_unset = arrangement.lastIndexOf(-1);

            int gaps = groups.size() + 1;
            int min_p_len = sum(groups) + gaps;
            int min_outside_space;
            if (first_unset == 0) {
                // no outer gaps needed
                min_p_len -= 2;
                min_outside_space = 0;
            }
            else {
                min_outside_space = 1;
            }

            boolean is_last = groups.size() <= 2;
            boolean is_single = groups.size() == 1;

            int used_before = sum(arrangement.subList(0, first_unset));
            int used_after = sum(arrangement.subList(last_unset + 1, arrangement.size()));
            List<Integer> remaining_springs = springs.subList(used_before, springs.size() - used_after);
            int free_len = remaining_springs.size() - min_p_len;
            if (free_len < 0) {
                return 0;
            }

            List<Integer> inner_groups;
            if (groups.size() > 2) {
                inner_groups = groups.subList(1, groups.size() - 1);
            }
            else {
                inner_groups = null;
            }

            long num_arrangements = 0;

            int max_s = min_outside_space + free_len + 1;
            for (int s = min_outside_space; s < max_s; s++) {
                if ( min_outside_space == 0 ) {
                    LocalDateTime currentDateTime = LocalDateTime.now();
                    String isoDateTime = currentDateTime.format(formatter);
                    int complete = Math.round((s - min_outside_space) / (float)max_s * 100);
                    System.out.println(isoDateTime + " " + getName() + " at " + complete + "%");
                }
                int springAt = remaining_springs.indexOf(1);
                if (springAt >= 0 && springAt < s) break;
                if (!testStart(remaining_springs, s, gs)) continue;

                int max_e = min_outside_space + free_len + 1 - (s - min_outside_space);
                for (int e = min_outside_space; e < max_e; e++) {
                    springAt = remaining_springs.lastIndexOf(1);
                    if (springAt >= remaining_springs.size() - e) break;
                    if (!testEnd(remaining_springs, e, ge)) continue;

                    if (is_last) {
                        if (is_single && (s + gs + e) < remaining_springs.size()) {
                            continue;
                        }
                        else {
                            springAt = remaining_springs.indexOf(1);
                            if (springAt >= (s + gs) && springAt <= remaining_springs.size() - (e + ge)) {
                                continue;
                            }
                        }
                    }

                    if (is_single || is_last || inner_groups == null) {
                        num_arrangements += 1;
                    }
                    else {
                        arrangement.set(first_unset, s);
                        arrangement.set(first_unset + 1, gs);
                        arrangement.set(last_unset - 1, ge);
                        arrangement.set(last_unset, e);
                        num_arrangements += arrange(arrangement, springs, inner_groups);
                        arrangement.set(first_unset, -1);
                        arrangement.set(first_unset + 1, -1);
                        arrangement.set(last_unset - 1, -1);
                        arrangement.set(last_unset, -1);
                    }
                }
            }
            return num_arrangements;
        }

        private static int sum(List<Integer> nums) {
            int sum = 0;
            for (int num : nums) {
                sum += num;
            }
            return sum;
        }

        private static boolean testStart(List<Integer> springs, int s, int gs) {
            for (int i = 0; i < s; i++) {
                if (springs.get(i) == 1) {
                    return false;
                }
            }
            for (int i = s; i < s + gs; i++) {
                if (springs.get(i) == 0) {
                    return false;
                }
            }
            return true;
        }

        private static boolean testEnd(List<Integer> springs, int e, int ge) {
            for (int i = 0; i < e; i++) {
                if (springs.get(springs.size() - i - 1) == 1) {
                    return false;
                }
            }
            for (int i = e; i < e + ge; i++) {
                if (springs.get(springs.size() - i - 1) == 0) {
                    return false;
                }
            }
            return true;
        }
    }
}
